compiler/luci/service/src/CircleShapeSignatureInferenceHelper.cpp

```diff
@@ -16,12 +16,130 @@
 
 #include "luci/Service/CircleShapeSignatureInferenceHelper.h"
 
+#include <loco.h>
+
+#include <luci/Log.h>
+
+#include <oops/InternalExn.h>
+
+namespace
+{
+
+luci::ShapeSignature clean_signature(const luci::ShapeSignature &signature)
+{
+  // If shape signature has at least one -1, it is not static.
+  for (uint32_t i = 0; i < signature.rank(); ++i)
+    if (signature.dim(i) == -1)
+      return signature;
+
+  // If all dimensions are static, return empty shape signature.
+  return luci::ShapeSignature();
+}
+
+} // namespace
+
 namespace luci
 {
 
 namespace ssinf
 {
 
+ShapeSignature reduced_signature(const loco::Node *node, const loco::Node *indices, bool keep_dims)
+{
+  ShapeSignature output_signature;
+
+  auto circle_node = loco::must_cast<const luci::CircleNode *>(node);
+  auto circle_indices = loco::must_cast<const luci::CircleNode *>(indices);
+  auto input_signature = circle_node->shape_signature();
+
+  // If shape signature is empty, output is static.
+  if (input_signature.rank() == 0)
+    return output_signature;
+
+  auto reduction_indices = dynamic_cast<const luci::CircleConst *>(circle_indices);
+
+  // When reduction_indices is not constant
+  if (reduction_indices == nullptr)
+  {
+    if (keep_dims)
+    {
+      // If keep_dims is true, rank is not changed.
+      output_signature.rank(input_signature.rank());
+      for (uint32_t i = 0; i < output_signature.rank(); ++i)
+        output_signature.dim(i) = -1;
+    }
+    else
+    {
+      // There is no way to inference for this case.
```

Comment:
maybe throw? or is it ok to return empty signature?

Reply:
Returning empty signature looked better for me..!

When TFLite converter cannot inference shape because of such cases, it returns no shape.
(The example is in https://github.com/Samsung/ONE/blob/master/res/TensorFlowPythonModels/examples/UnknownExpandDims/__init__.py)
And `luci` is supporting `ShapeStatus::NOSHAPE` for those case.

However, as I am not sure that returning empty shape signature is really correct, just throwing is also not bad :)

Reply:
as me may not be certain for this case, how about adding `assert` just for debugging?

Reply:
with a comment 
```cxx
// let's break with debug version just to find this case
``